internal/sync2/pipe.go

```diff
@@ -0,0 +1,280 @@
+	pipe := writer.pipe
+	pipe.mu.Lock()
+
+	// has the reader finished?
+	if pipe.readerDone {
+		pipe.mu.Unlock()
+		return 0, pipe.readerErr
+	}
+
+	// have we closed already
+	if pipe.writerDone {
+		pipe.mu.Unlock()
+		return 0, io.ErrClosedPipe
+	}
+
+	// check how much do they want to write
+	canWrite := pipe.limit - pipe.write
+
+	// no more room to write
+	if canWrite == 0 {
+		pipe.mu.Unlock()
+		return 0, io.ErrClosedPipe
+	}
+
+	// figure out how much to write
+	toWrite := int64(len(data))
+	if toWrite > canWrite {
+		toWrite = canWrite
+	}
+
+	writeAt := pipe.write
+	pipe.mu.Unlock()
+
+	// write data to buffer
+	writeAmount, err := pipe.buffer.WriteAt(data[:toWrite], writeAt)
+
+	pipe.mu.Lock()
+	// update writing head
+	pipe.write += int64(writeAmount)
+	// wake up reader
+	pipe.nodata.Broadcast()
+	// check whether we have finished
+	done := pipe.write >= pipe.limit
+	pipe.mu.Unlock()
+
+	if err == nil && done {
+		err = io.ErrClosedPipe
+	}
+	return writeAmount, err
+}
+
+// Read reads from the pipe returning io.EOF when writer is closed or blockSize is reached
+func (reader pipeReader) Read(data []byte) (n int, err error) {
+	pipe := reader.pipe
+	pipe.mu.Lock()
+	// wait until we have something to read
+	for pipe.read >= pipe.write {
+		// has the writer finished?
+		if pipe.writerDone {
+			pipe.mu.Unlock()
+			return 0, pipe.writerErr
+		}
+
+		// have we closed already
+		if pipe.readerDone {
+			pipe.mu.Unlock()
+			return 0, io.ErrClosedPipe
+		}
+
+		// have we run out of the limit
+		if pipe.read >= pipe.limit {
+			pipe.mu.Unlock()
+			return 0, io.EOF
+		}
+
+		// ok, lets wait
+		pipe.nodata.Wait()
+	}
+
+	// how much there's available for reading
+	canRead := pipe.write - pipe.read
+	// how much do they want to read?
+	toRead := int64(len(data))
+	if toRead > canRead {
+		toRead = canRead
+	}
+	readAt := pipe.read
+	pipe.mu.Unlock()
+
+	// read data
+	readAmount, err := pipe.buffer.ReadAt(data[:toRead], readAt)
+
+	pipe.mu.Lock()
+	// update info on how much we have read
+	pipe.read += int64(readAmount)
+	done := pipe.read >= pipe.limit
+	pipe.mu.Unlock()
+
+	if err == nil && done {
+		err = io.EOF
+	}
+	return readAmount, err
+}
+
+// MultiPipe is a multipipe backed by a single file
+type MultiPipe struct {
+	pipes    []pipe
+	refcount int64
+}
+
+// NewMultiPipeFile returns a new MultiPipe that is created in tempdir
+// if tempdir == "" the fill will be created it into os.TempDir
+func NewMultiPipeFile(tempdir string, blockCount, blockSize int64) (*MultiPipe, error) {
+	tempfile, err := ioutil.TempFile(tempdir, "multifilepipe")
+	if err != nil {
+		return nil, err
+	}
+
+	err = tempfile.Truncate(blockCount * blockSize)
+	if err != nil {
+		closeErr := tempfile.Close()
+		if closeErr != nil {
+			return nil, fmt.Errorf("%v/%v", err, closeErr)
+		}
+		return nil, err
+	}
+
+	multipipe := &MultiPipe{
+		pipes:    make([]pipe, blockCount),
+		refcount: blockCount,
```

Comment:
@egonelbre Shouldn't this be `2 * blockCount`, i.e. 2 halves for each pipe?

Reply:
No, because the refcount counts open pipes. So `pipe` has it's own refcount for tracking either side of the pipe. i.e. you need two closes in `pipe` for `multipipe.refcount` to be decreased.

Reply:
Oh, yes. I am confusing the different refcounts.